### PR TITLE
Add MTD driver function for iMXRT

### PR DESCRIPTION
**Pull request by HarrisonChoi750405**

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [827d052d1821a5ad1de3a25cd69fd9dc56916a7e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [827d052d1821a5ad1de3a25cd69fd9dc56916a7e] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+
+	fdbg("priv: 0x%x,  IMXRT_FLEXSPI: 0x%x,  offset: 0x%x\n", priv, IMXRT_FLEXSPI, (long)offset);
+
+	status = imxrt_flexspi_nor_flash_erase_sector(IMXRT_FLEXSPI, offset);
+	if (status != kStatus_Success) {
+		fdbg("Erased fail!\n");
+		return;
+	}
+
+	fdbg("Erased\n");
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: This message shows information, not errors.
So this needs to be 'fvdbg'.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+	(void)SPI_SEND(priv->dev, IMXRT_BE);
+
+	/* Deselect the FLASH */
+
+	SPI_SELECT(priv->dev, SPIDEV_FLASH, false);
+	
+	size_t sectors_erased = 0;
+	size_t startsector = 8;
+//	fdbg("startsector: %d nsectors: %d\n", (long)startsector, (int)priv->nsectors);
+	fdbg("startsector: %d nsectors: %d\n", (long)startsector, 2);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: I'll fix this line in my PR (to be submitted today).

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+	/* Send the "Bulk Erase (BE)" instruction */
+
+	(void)SPI_SEND(priv->dev, IMXRT_BE);
+
+	/* Deselect the FLASH */
+
+	SPI_SELECT(priv->dev, SPIDEV_FLASH, false);
+	
+	size_t sectors_erased = 0;
+	size_t startsector = 8;
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: I'll fix this line in my PR (to be submitted today).

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+//	fdbg("startsector: %d nsectors: %d\n", (long)startsector, (int)priv->nsectors);
+	fdbg("startsector: %d nsectors: %d\n", (long)startsector, 2);
+
+	/* Lock access to the SPI bus until we complete the erase */
+
+//	while (sectors_erased < priv->nsectors) {
+	while (sectors_erased < 2) {
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: I'll fix this line in my PR (to be submitted today).

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+		/* Not using sub-sector erase.  Erase each full sector */
+
+		imxrt_sectorerase(priv, startsector + sectors_erased);
+		sectors_erased++;
+	}
+
+	fdbg("Return = erased sectors = %d\n", sectors_erased);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+	return sectors_erased;
+	#endif
+
+	fdbg("Return: OK\n");
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fvdbg

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+}
+
+/************************************************************************************
+ * Name: imxrt_read
+ ************************************************************************************/
+
+static ssize_t imxrt_read(FAR struct mtd_dev_s *dev, off_t offset, size_t nbytes, FAR uint8_t *buffer)
+{
+	status_t status;
+	//	FAR struct imxrt_dev_s *priv = (FAR struct imxrt_dev_s *)dev;
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: Let's delete this if unnecessary..

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+	fvdbg("offset: %08lx nbytes: %d\n", (long)offset, (int)nbytes);
+
+	/* We must test if the offset + count crosses one or more pages
+	 * and perform individual writes.  The devices can only write in
+	 * page increments.
+	 */
+
+	startpage = offset / (1 << priv->pageshift);
+	endpage = (offset + nbytes) / (1 << priv->pageshift);
+//	startpage = offset / (1 << priv->sectorshift);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: Need to delete old comments.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+//	endpage = (offset + nbytes) / (1 << priv->sectorshift);
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: Need to delete old comments.

File: os/arch/arm/src/imxrt/imxrt_norflash.c [hunk 10]
```diff
@@ -839,6 +725,125 @@ static inline void flexspi_clock_init(void)
...
+ ************************************************************************************/
+status_t imxrt_flexspi_nor_flash_page_program(FLEXSPI_Type *base, uint32_t address, const uint32_t *src)
+{
+	status_t status;
+	flexspi_transfer_t flashXfer;
+	irqstate_t flags;
+
+	flags = irqsave();
+
+	/* Write neable */
```
Comment (chanijjani) [inline os/arch/arm/src/imxrt/imxrt_norflash.c]: enable...

Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_norflash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 9 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: removed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 8 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: removed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 7 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: removed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c (hunk 6 above)
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+	#ifdef CONFIG_IMXRT_SUBSECTOR_ERASE
+	if (priv->subsectorshift > 0) {
+		offset = sector << priv->subsectorshift;
+	} else
+	#endif
+	{
+		offset = sector << priv->sectorshift;
+	}
+
+	fdbg("priv: 0x%x,  IMXRT_FLEXSPI: 0x%x,  offset: 0x%x\n", priv, IMXRT_FLEXSPI, (long)offset);
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

**Comment by seinfra** [comment]
Target : [1c8572f7ef55fb3c941b1781a84e53bcef901152] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [1c8572f7ef55fb3c941b1781a84e53bcef901152] - Code Rule Check (C++) OK.

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+
+/* Helpers */
+static inline int imxrt_readid(struct imxrt_dev_s *priv);
+static void imxrt_waitwritecomplete(struct imxrt_dev_s *priv);
+static inline void imxrt_sectorerase(struct imxrt_dev_s *priv, off_t offset);
+static inline int imxrt_bulkerase(struct imxrt_dev_s *priv, unsigned long arg);
+static inline void imxrt_pagewrite(struct imxrt_dev_s *priv, FAR const uint8_t *buffer, off_t offset);
+
+/* MTD driver methods */
+
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_mtd_flash.c [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+static void imxrt_waitwritecomplete(struct imxrt_dev_s *priv)
+{
+	status_t status;
+
+	status = imxrt_flexspi_nor_wait_bus_busy(IMXRT_FLEXSPI);
+
+	if (status != kStatus_Success) {
+		fdbg("imxrt_flexspi_nor_wait_bus_busy: error[%d]\n", status);
+	}
+
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_mtd_flash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_hyperflash.c [hunk 14]
```diff
@@ -478,12 +395,284 @@ static inline void flexspi_clock_init(void)
...
+ ************************************************************************************/
+status_t imxrt_flexspi_nor_flash_page_program(FLEXSPI_Type *base, uint32_t address, const uint32_t *src)
+{
+	status_t status;
+	flexspi_transfer_t flashXfer;
+	irqstate_t flags;
+
+	flags = irqsave();
+
+	/* Write neable */
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_hyperflash.c]: fixed

File: os/arch/arm/src/imxrt/imxrt_hyperflash.c [hunk 15]
```diff
@@ -478,12 +395,284 @@ static inline void flexspi_clock_init(void)
...
+ * Description:
+ *   Write data to hyper flash using flexspi interface
+ *
+ ************************************************************************************/
+status_t imxrt_flexspi_nor_write_enable(FLEXSPI_Type *base, uint32_t baseAddr)
+{
+	flexspi_transfer_t flashXfer;
+	status_t status;
+
+	/* Write neable */
```
[the remark this answers is not in the document]
Comment (HarrisonChoi750405) [reply os/arch/arm/src/imxrt/imxrt_hyperflash.c]: fixed

**Comment by seinfra** [comment]
Target : [bcdeb6774ace501aed7b826728a2412aff47b716] -

**Comment by seinfra** [comment]
Target : [bcdeb6774ace501aed7b826728a2412aff47b716] -

**Comment by chanijjani** [comment]
@HarrisonChoi750405 And please add some descriptions of this PR.

**Comment by seinfra** [comment]
Target : [e6edb12417313a34deef04d917ce2e3cbbee871a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e6edb12417313a34deef04d917ce2e3cbbee871a] - Code Rule Check OK.

**Pull request merged**